server/data/data-library/src/main/java/org/apache/james/domainlist/lib/AbstractDomainList.java

```diff
@@ -151,32 +151,33 @@ public boolean containsDomain(String domain) throws DomainListException {
     }
 
     @Override
-    public List<String> getDomains() throws DomainListException {
+    public ImmutableList<String> getDomains() throws DomainListException {
```

Comment:
How about:

```
    @Override
    public List<String> getDomains() throws DomainListException {
        ImmutableList<String> domainsWithoutIp = Stream.concat(
            getDomainListInternal().stream(),
            detectDomains().stream())
            .collect(Guavate.toImmutableList());
        List<String> detectedIps = detectIps(domainsWithoutIp);
        
        return Stream.concat(
            domainsWithoutIp.stream(),
            detectedIps.stream())
            .peek(domain ->  LOGGER.debug("Handling mail for {}" + domain))
            .collect(Guavate.toImmutableList());
    }
```

?

Reply:
I think that this has the same issue as the current code in that it creates temporary `ImmutableList` instances, which, in order to be concatenated, need to be copied to a larger `ImmutableList` that stores the result of concatenation.

Another source of temporary `ImmutableList` instances is getDomainsIP()/getDomainIP(), which currently returns a new `ImmutableList` for each domain. I have pushed a commit to refactor these private utility methods so that they work with `Stream`s instead. This should be more efficient, I think.